File: scripts/bindings/shape_dashes.py
"""FIX-S09 curve dash gaps, exact source identities and retained host output."""
from pathlib import Path
import sys,json
ROOT=Path(__file__).resolve().parents[2];sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(100.,100.).dpi(72).layout(c.layout_options().padding(0));keys=[9007199254740993,9007199254740997]
data=c.Data.columns({'x':[0.,1.],'y':[.5,.5]},keys=keys)
p=c.plot(data).aes(c.aes().x('x').y('y')).layer(c.shape_line().curve({'kind':'BumpX'}).style(c.style().dashes([10.,10.]))).x_axis(c.x_axis().scale(c.scale_linear().domain(0.,1.)).visible(False)).y_axis(c.y_axis().scale(c.scale_linear().domain(0.,1.)).visible(False)).build()
wire=p.to_json();q=c.Plot.from_json(wire);assert q.to_json()==wire;q.dispose();chart=p.chart();p.dispose();f=chart.present(output,options);scene=f.scene();i=next(i for i,item in enumerate(scene['items'])if'ShapePath'in item['primitive']);shape=scene['items'][i]['primitive']['ShapePath'];assert shape['dashes']==[10.,10.];assert 'CubicTo'in shape['geometry']['commands'][1];assert [int(t['Source']['key'])for t in scene['targets'][i]]==keys
for x in [5.,25.,45.,65.,85.]:assert len(chart.inspect(x,50.,mode='Containment')['targets'])==1
for x in [15.,35.,55.,75.,95.]:assert chart.inspect(x,50.,mode='Containment')['targets']==[]
assert chart.inspect(-1.,50.,mode='Containment')['targets']==[]
saved={fmt:f.export(fmt)for fmt in ['svg','pdf','png']};assert b'stroke-dasharray="10,10"'in saved['svg'];chart.dispose()
for fmt,expected in saved.items():assert f.export(fmt)==expected
f.dispose();data.dispose();output.dispose();print('PASS Python retained curved dashes, five ink hits/five gap misses, clipping, exact source keys, wire round trip and exports after disposal.')

File: crates/chart-export/tests/shape_dashes.rs
//! FIX-S09: curved and filled vector strokes retain dash styling in immutable output.
use chart_core::plot::shape_line;
use chart_core::{
    composition::Anchor,
    path::Path,
    prelude::*,
    scene::{Color, Primitive, Stroke},
    shape::CurveSpec,
};
use chart_export::*;
const FONT: &[u8] = include_bytes!("../../../fixtures/capability/fonts/NotoSans-Regular.ttf");
#[test]
fn curved_svg_and_pdf_png_retain_dash_pattern_fill_and_snapshot() {
    let data = Data::columns()
        .column("x", [0., 1.])
        .column("y", [0., 1.])
        .build()
        .unwrap();
    let mut path = Path::default();
    path.arc([0., 0.], 20., 0., std::f64::consts::TAU, false)
        .unwrap();
    path.move_to(10., 0.).unwrap();
    path.arc([0., 0.], 10., 0., std::f64::consts::TAU, true)
        .unwrap();
    let black = Color {
        red: 0,
        green: 0,
        blue: 0,
        alpha: 255,
    };
    let yellow = Color {
        red: 230,
        green: 200,
        blue: 30,
        alpha: 255,
    };
    let make = |pattern| {
        plot(data.clone())
            .aes(aes().x("x").y("y"))
            .layer(shape_line().curve(CurveSpec::BumpX))
            .layer(
                vector_path("ring", path.geometry())
                    .anchor(Anchor::Output { x: 150., y: 80. })
                    .fill(Some(yellow))
                    .stroke(Some(Stroke {
                        color: black,
                        width: 2.,
                    })),
            )
            .theme(theme().style(style().dashes(pattern)))
            .build()
            .unwrap()
    };
    let output = Output::new(FONT).unwrap();
    let options = export_options(PageSize::points(300., 160.).unwrap()).dpi(72);
    let solid = output
        .request(&make(vec![]), options.clone())
        .unwrap()
        .prepare()
        .unwrap();
    let retained = output.request(&make(vec![5., 3.]), options).unwrap();
    let dashed = retained.prepare().unwrap();
    assert_eq!(dashed.scene().wire_version(), 4);
    let vector = dashed
        .scene()
        .items()
        .iter()
        .find_map(|i| {
            if let Primitive::VectorPath {
                geometry,
                fill,
                stroke,
                dashes,
            } = &i.primitive
            {
                Some((geometry, fill, stroke, dashes))
            } else {
                None
            }
        })
        .unwrap();
    assert_eq!(*vector.1, Some(yellow));
    assert_eq!(
        *vector.2,
        Some(Stroke {
            color: black,
            width: 2.
        })
    );
    assert_eq!(vector.3, &vec![5., 3.]);
    assert_eq!(vector.0.commands().len(), path.geometry().commands().len());
    let svg = String::from_utf8(dashed.export(Format::Svg).unwrap().bytes).unwrap();
    assert_eq!(svg.matches("stroke-dasharray=\"5,3\"").count(), 2);
    assert!(svg.contains("fill-rule=\"nonzero\""));
    for format in [Format::Svg, Format::Pdf, Format::Png] {
        let bytes = dashed.export(format).unwrap().bytes;
        assert_ne!(bytes, solid.export(format).unwrap().bytes);
        assert_eq!(
            bytes,
            retained.prepare().unwrap().export(format).unwrap().bytes
        );
    }
}
